x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/mapreduce/MapReduceValueSourceRegistry.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.aggs.mapreduce;
+
+import org.elasticsearch.search.aggregations.support.CoreValuesSourceType;
+import org.elasticsearch.search.aggregations.support.ValuesSourceConfig;
+import org.elasticsearch.search.aggregations.support.ValuesSourceRegistry;
+
+import java.util.List;
+
+public final class MapReduceValueSourceRegistry {
```

Comment:
I agree with Ben.  If the Map Reduce framework is meant to be general, why are we forcing every client to use the ValuesSourceRegistry in the same way?  And if it's not meant to be generic, what is the point of it?

Reply:
I disagree, it is IMO a strength not to work with internal details. A map reducer takes key value pairs, and that what this implementation provides, it translates lucene internals and produces proper typed _generic_ values. Why should this be differently extracted?

As written in the other comment, this is heavily inspired by multi terms. It seems to me that all new aggs should use the value source registry if possible. If that's not the case, I can revert back to an own implementation.

> why are we forcing every client to use the ValuesSourceRegistry in the same way?

I think we don't know the answer yet. Currently I don't see a reason that it should be different and therefore I wouldn't implement an extra abstraction layer. If we realize an abstraction is needed it can be added later.

Reply:
I see.  I think the problem is that, to me, this looks like an abstraction layer.  I'm hearing you say that it's not, that the whole MapReduce package is tightly coupled to the FrequentTerms aggregation, and not meant to be generic?  Is that correct?

Because this makes more sense if it's only applying to a single aggregation rather than as part of a new framework.

Having said that, if this is not meant to be a new abstraction, I don't understand why it's being pulled out into an independent package that looks a lot like a new abstraction.  Why not put the ValuesSourceRegistry stuff in FrequentItemSetsAggregationBuilder, which is where I'd expect to find it in any other aggregation?